### ISSUE TITLE
`dotnet run` of a newly created ASP.NET Core project on MacOS prompts for password to allow keychain access

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/sdk#22544]
Feedback from https://twitter.com/leastprivilege/status/1458505534093643776

## Problem
When using `dotnet run` on a newly created ASP.NET Core project on MacOS, the user is prompted by the OS to allow access to the Keychain so that the app can access the HTTPS developer certificate. After entering their password and clicking "Allow" or "Always Allow" the app starts. This process must be repeated for every new ASP.NET Core project created.

*MacOS prompt showing message "Host6 wants to access key "tmpE4viKq" in your keychain."
A textbox for the password and three buttons are also shown: Always Allow, Deny, Allow.*

![image](https://user-images.githubusercontent.com/249088/141239495-4b526dae-0342-4d17-90d3-210e29d7dc44.png)

## Details
In .NET 6 `dotnet run` was changed to launch the app via the AppHost executable rather than via `dotnet MyApp.dll`. While this has advantages (e.g. support for cross-architecture app launching, the process name being the project name, etc.) it results in every new project being a new executable from the OS's security point-of-view, and thus requires each new web app to explicitly be granted access to the Keychain.

Note the issue doesn't occur when launching via Visual Studio for Mac 2022 as it doesn't use `dotnet run` to launch the app.

## Workaround
To workaround this issue the [`<UseAppHost>` MSBuild property](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#useapphost) can be set in the project file to `false`, e.g.:

```xml
<PropertyGroup>
    <UseAppHost>false</UseAppHost>
</PropertyGroup>
```

You can also pass this property as part of the `dotnet run` command itself:

```command
> dotnet run -p:UseAppHost=false
```

Note that this property completely disables a native executable being created for the project during build, not just in the context of launching the project via `dotnet run`.

## Proposed Changes
We should ensure users are not faced with this prompt as part of the default experience for creating and running ASP.NET Core apps on MacOS.

To that end, we could consider changing the default value for the `<UseAppHost>` MSBuild property to `false` when a project uses the `Microsoft.NET.Sdk.Web` SDK. As noted above however this would actually disable the production of a native executable during build, and so has side-effects beyond the scope of launching the project via `dotnet run`. For that reason we should likely consider alternate approaches:
- We could consider adding support to `dotnet run` for controlling this behavior via a new environment variable, e.g. `DOTNET_USE_APP_HOST`, such that it could be set in the scope of the machine, profile, session, or process launch.
- We could also consider adding a new property to `launchSettings.json` under the `"profiles"` section, placed on a profile, that controls whether the project is launched via the AppHost or not when launched from `dotnet run` or an IDE (i.e. this property should also be honored in environments that launch the project using the AppHost today, like Visual Studio).
  - Note that while `launchSettings.json` supports setting environment variables for the launched app process, attempting to use that to set the variable proposed above (`DOTNET_USE_APP_HOST`) would not work as that variable needs to be set in the context of `dotnet run` itself, not the app subsequently being launched).

**Comment by khalidabuhakmeh** [comment]
I'm adding this documentation for folks to better understand the issue with `AppHost` and `macOS`. https://docs.microsoft.com/en-us/dotnet/core/install/macos-notarization-issues#apphost-is-disabled-by-default

**Comment by agocke** [comment]
> To that end, we could consider changing the default value for the <UseAppHost> MSBuild property to false when a project uses the Microsoft.NET.Sdk.Web SDK

To be clear, this will break launching .NET 5 and .NET 3.1 apps on multi-arch (M1) machines.

**Comment by DamianEdwards** [comment]
@agocke right, that's one of the benefits of the AppHost (as I call out). This consideration might be a reason to ensure that if we add a new `launchSettings.json` flag it's done in such a way that it either only applies when there isn't a cross-arch scenario to deal with, or that it fails with a meaningful message that instructs the user on how to configure the launch to support it.

**Comment by dsplaisted** [comment]
FYI @mateoatr @vitek-karas, looks like https://github.com/dotnet/docs/issues/24766 impacts ASP.NET apps using HTTPs, each build is considered a separate app from MacOS perspective, so you have to enter the password each time.

**Comment by DamianEdwards** [comment]
It seems the [doc about macOS Notarization](https://docs.microsoft.com/en-us/dotnet/core/install/macos-notarization-issues#apphost-is-disabled-by-default) is still [incorrectly saying that the app host is disabled by default](https://docs.microsoft.com/en-us/dotnet/core/install/macos-notarization-issues#apphost-is-disabled-by-default) for projects on macOS? This is incorrect since .NET 6.0 isn't it?

**Comment by agocke** [comment]
Yup, that looks wrong. I'll make a PR.

**Comment by agocke** [comment]
Doc fix https://github.com/dotnet/docs/pull/27984

**Comment by DamianEdwards** [comment]
@marcpopMSFT  Are we going to look at addressing this issue? I note this is in the discussion milestone now but I think there is work here to improve the default experience for ASP.NET Core dev on macOS. As noted, if we simply disable the apphost in new ASP.NET Core apps or the Web SDK, the benefits of apphost could be lost (e.g. cross arch launching), so the assumption is a change would have to be made in `dotnet run`. Or is there an acceptable way to achieve the same thing in the web SDK itself?

**Comment by agocke** [comment]
Just curious: how do languages like Go deal with this? It seems like authorizing individual apps is the expected flow from the MacOS perspective, so it feels like we're swimming against the tide in some sense.

**Comment by DamianEdwards** [comment]
@agocke that's an interesting question, but I'm not sure Go is as popular for building the kinds of web site apps that folks do with ASP.NET Core, nodeJS, Python, etc. I'm happy to discuss the solution here being something that just the web SDK opts in to by default.

**Comment by DamianEdwards** [comment]
pinging some others from ASP.NET for awareness @danroth27 @LadyNaggaga @captainsafia

**Comment by marcpopMSFT** [comment]
> @marcpopMSFT Are we going to look at addressing this issue? 
@DamianEdwards that's a question for Andy as it's connected to the host creation so either don't use the apphost for Web projects unless there's a way to avoid the keychain check somehow that we don't know of.

**Comment by DamianEdwards** [comment]
@marcpopMSFT my understanding of this issue leads me to the following potential options for resolution:
1. Change the way `dotnet run` works so that it does **not** launch the project via the app host in certain scenarios, perhaps driven by a `launchSettings.json` property. So the defaults about app host generation would remain, but the way `dotnet run` launches the project would differ. This approach means we could potentially retain the benefits of the app host for the scenarios where it might be preferred, as I [commented here](https://github.com/dotnet/sdk/issues/22544#issuecomment-966535551).
2. Change something in the web SDK so that web projects default to not generating an app host on macOS. This would mean users will get issues in scenarios where the app host is beneficial, like cross-architecture.

The first option seems like it would be a better compromise and that requires updating `dotnet run` in the CLI doesn't it?

**Comment by marcpopMSFT** [comment]
@DamianEdwards , are you suggesting adding a new launch settings flag and would we then essentially be requiring that file for all Mac users?

**Comment by DamianEdwards** [comment]
@marcpopMSFT nothing would be requiring the file, but if it's present (like in our templates), and the property is set, then it would affect how the app is launched. If the property is not set the behavior would remain as it is now. As @agocke pointed out, there are advantages to the apphost in cross-architecture scenarios so the suggestion is for the setting to enable use of the apphost *only* when it would benefit.

**Comment by DamianEdwards** [comment]
@marcpopMSFT @baronfel this is still tripping up folks today on macOS (including @robconery just now). Can we get this prioritized for preview.6 or preview.7?

**Comment by khalidabuhakmeh** [comment]
Yes please :)

**Comment by chrisfcarroll** [comment]
As a lowly cross-platform end-developer of webservices, I concur that suggestions so far for using a property in either the msbuild `.csproj` file or in `launchSettings.json`, pre-set in the templates, are about the least-surprising places I might think to look for some platform-specific stuff.

The current behaviour —the KeyChain popup coming up— would be okay if the CLI printed information lines telling me how to avoid it?

```
    <!-- Remove this line if you do not use OSX. You can also remove it on OSX if you grant this project access to the KeyChain item "tmpfRiltr" on each machine that it runs on. See https://github.com/dotnet/sdk/issues/22544 -->
    <UseAppHost Condition="$([MSBuild]::IsOSPlatform('OSX'))">false</UseAppHost>
```

**Comment by DamianEdwards** [comment]
@baronfel based on the comments up to this point, I think it might be better to have a property that only impacts the launch behavior, not what gets built. Today, VS for Mac and VS Code (by default) still launch the app using `dotnet MyWebApp.dll` rather than the app host (which avoids the details of this issue) but CLI launches via `dotnet run` and `dotnet watch` are impacted. Properties in `launchSettings.json` shouldn't impact built artifacts IMO.

Updated proposal: introduce a new property that can be specified on profiles in `launchSettings.json`, `PreferAppHost`, that defaults to `true`, and update the web project templates to set this property to `false` when creating a new project on macOS. As you point out, we should ensure all launch environments including VS, VS for Mac, VS Code, `dotnet run`, and `dotnet watch` honor this setting when launching the project.

**Comment by baronfel** [comment]
@DamianEdwards  I think we were saying the same thing. I was only ever meaning an evaluation-time property modification, never a build-time one. I've got a WIP implementation on the linked PR, and we'll work on getting this good to go 👍

**Comment by DamianEdwards** [comment]
Note we're investigating making changes on the ASP.NET Core side that would mitigate this issue too, specifically changing Kestrel to get the dev cert from a disk location rather than from the Keychain. This would avoid the prompt for each new web application. I still need to do some verification around whether client apps (using `HttpClient`) cause a prompt when they verify the certificate is trusted via the Keychain.

**Comment by khalidabuhakmeh** [comment]
Yeah... disk certs would be cool, but it wouldn't be trusted by browsers without extra steps. *cough* LINUX *cough*

**Comment by DamianEdwards** [comment]
@khalidabuhakmeh that sounds like a great proposal for a separate issue and community-implemented prototype :smile:

**Comment by Pilchie** [comment]
I think @davidfowl was recently talking about building something similar to @khalidabuhakmeh's proposal with YARP - basically a single proxy on a machine that other apps could discover and register themselves with.

**Comment by davidfowl** [comment]
> It could also manage custom TLDs specific to apps.

@khalidabuhakmeh how would that work?

**Comment by khalidabuhakmeh** [comment]
@davidfowl yes, it would manage entries in the host file of the machine.

**Comment by davidfowl** [comment]
Yea, that doesn't sound great 😄

**Comment by khalidabuhakmeh** [comment]
I mean... dotnet already asks you for administrative privileges to jam a certificate into the keychain. So if security is your concern, well... I think there are more fruitful attack vectors. 😅

**Comment by DamianEdwards** [comment]
> I mean... dotnet already asks you for administrative privileges to jam a certificate into the keychain. So if security is your concern, well... I think there are more fruitful attack vectors. 

Technically, it's not adding the cert that requires the elevation, it's marking it as a trusted root authority in the system keychain. We're planning on [making changes to that in .NET 7 to narrow the scope of install](https://github.com/dotnet/aspnetcore/issues/41878) and trust to just the user's login keychain which should reduce the prompting to just a single one when trusted.

**Comment by khalidabuhakmeh** [comment]
https://github.com/aspnet/Announcements/issues/486 I guess this solves the problem. 😅

**Comment by DamianEdwards** [comment]
> https://github.com/aspnet/Announcements/issues/486 I guess this solves the problem. 😅

Yep! That's the idea.

**Comment by khalidabuhakmeh** [comment]
It does make me wonder what the best practice around certificates will be moving forward for the #aspnetcore developer community.

Will it become more in line with what other communities do, where there's always assumed to be some kind of proxy that handles certificates and HTTPS traffic?

If so, a proxy project template with a starter cert might be a good addition to the web templates already included in .NET.

**Comment by davidfowl** [comment]
@khalidabuhakmeh We don't make assumptions about your deployment environment. When people ask "what is recommended" then we need to answer with another question, because if we said, "to run in Kubernetes like this", then maybe that doesn't fit your deployment model. 

What you ask needs to be caveated with "when running in this environment". Now, if you're asking what the recommended environment is then......You're asking the wrong people.

**Comment by khalidabuhakmeh** [comment]
Yep, totally understand. Context matters I'm not going to argue with that. :)

**Comment by DamianEdwards** [comment]
I think the specific issue we need completed to resolve this experience problem on macOS is dotnet/aspnetcore#41879. Once that change is done, I think we should close this issue.

**Comment by DamianEdwards** [comment]
Closing this now that dotnet/aspnetcore#41879 is completed.

**Comment by chrisfcarroll** [comment]
For this, and so many things, Thankyou!

**Comment by dluc** [comment]
I suspect this is related to consecutive builds of the same C# app losing access to MacOS keychain in dev workstations, failing to retrieve AAD secrets. 

For instance when using `InteractiveBrowserCredential` the AAD library caches encrypted tokens info in the keychain (without prompting the user). When developing an app, changing the code and rebuilding after a few minutes from the cache write, a new build is unable to read the cache written by the previous build, crashing with errors like this (note `KeyChain authorization/authentication failed. .Error code: -25293. OS error code -25293.`):

```
Azure.Identity.AuthenticationFailedException: InteractiveBrowserCredential authentication failed: Persistence check failed. Reason: KeyChain authorization/authentication failed. .Error code: -25293. OS error code -25293.

 ---> Microsoft.Identity.Client.Extensions.Msal.MsalCachePersistenceException: Persistence check failed. Reason: KeyChain authorization/authentication failed. .Error code: -25293. OS error code -25293.

 ---> Microsoft.Identity.Extensions.InteropException: KeyChain authorization/authentication failed. .Error code: -25293
   at Microsoft.Identity.Extensions.Mac.SecurityFramework.ThrowIfError(Int32 error, String defaultErrorMessage)
   at Microsoft.Identity.Client.Extensions.Msal.MacOSKeychain.Get(String service, String account)
   at Microsoft.Identity.Client.Extensions.Msal.MacKeychainAccessor.Read()
   at Microsoft.Identity.Client.Extensions.Msal.Storage.VerifyPersistence()
   --- End of inner exception stack trace ---

   at Microsoft.Identity.Client.Extensions.Msal.Storage.VerifyPersistence()
   at Microsoft.Identity.Client.Extensions.Msal.MsalCacheHelper.VerifyPersistence()
   at Azure.Identity.MsalCacheHelperWrapper.VerifyPersistence()
   at Azure.Identity.TokenCache.GetCacheHelperAsync(Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.TokenCache.RegisterCache(Boolean async, ITokenCache tokenCache, CancellationToken cancellationToken)
   at Azure.Identity.MsalClientBase`1.GetClientAsync(Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.MsalPublicClient.AcquireTokenSilentCoreAsync(String[] scopes, String claims, AuthenticationRecord record, String tenantId, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.MsalPublicClient.AcquireTokenSilentAsync(String[] scopes, String claims, AuthenticationRecord record, String tenantId, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.InteractiveBrowserCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---

   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex, String additionalMessage)
   at Azure.Identity.InteractiveBrowserCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.InteractiveBrowserCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
```

**Comment by DamianEdwards** [comment]
@dluc might be worth logging an issue on the AAD library to see if there's something they can do to improve the dev experience there.
